src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteExecutorKeepalived.java

```diff
@@ -0,0 +1,308 @@
+      this.callCredentialsProvider = callCredentialsProvider;
+      this.backoff = new ProgressiveBackoff(this.retrier::newBackoff);
+      this.executionBlockingStubSupplier = executionBlockingStubSupplier;
+    }
+
+    ExecuteResponse start() throws IOException, InterruptedException {
+      // Execute has two components: the Execute call and (optionally) the WaitExecution call.
+      // This is the simple flow without any errors:
+      //
+      // - A call to Execute returns streamed updates on an Operation object.
+      // - We wait until the Operation is finished.
+      //
+      // Error possibilities:
+      // - An Execute call may fail with a retriable error (raise a StatusRuntimeException).
+      //   - If the failure occurred before the first Operation is returned and tells us the
+      //     execution is accepted, we retry the call.
+      //   - Otherwise, we call WaitExecution on the Operation.
+      // - A WaitExecution call may fail with a retriable error (raise a StatusRuntimeException).
+      //   In that case, we retry the WaitExecution call on the same operation object.
+      // - A WaitExecution call may fail with a NOT_FOUND error (raise a StatusRuntimeException).
+      //   That means the Operation was lost on the server, and we will retry to Execute.
+      // - Any call can return an Operation object with an error status in the result. Such
+      //   Operations are completed and failed; however, some of these errors may be retriable.
+      //   These errors should trigger a retry of the Execute call, resulting in a new Operation.
+      Preconditions.checkState(lastOperation == null);
+
+      ExecuteResponse response = null;
+      // Exit the loop as long as we get a response from either Execute() or WaitExecution().
+      while (response == null) {
+        // We use refreshIfUnauthenticated inside retry block. If use it outside, retrier will stop
+        // retrying when received a unauthenticated error, and propagate to refreshIfUnauthenticated
+        // which will then call retrier again. It will reset the retry time counter so we could
+        // retry more than --remote_retry times which is not expected.
+        response = retrier
+            .execute(() -> Utils.refreshIfUnauthenticated(this::execute, callCredentialsProvider),
+                backoff);
+
+        // If no response from Execute(), use WaitExecution() in a "loop" which is implicated inside
+        // the retry block.
+        if (response == null) {
+          response = retrier.execute(
+              () -> Utils.refreshIfUnauthenticated(this::waitExecution, callCredentialsProvider),
+              backoff);
+        }
+      }
+
+      return response;
+    }
+
+    @Nullable
+    ExecuteResponse execute() throws IOException {
+      Preconditions.checkState(lastOperation == null);
+
+      try {
+        Iterator<Operation> streamOperations = executionBlockingStubSupplier.get().execute(request);
```

Comment:
nit: call the variables `operations` or `operationStream`? (the way I parse "streamOperations" is that it "streams Operations", that is, an activity which is better suited as a name for a method and "operationStream" sounds like "a stream of Operations" to me, which is a noun and is thus a better variable name)

Reply:
Done.